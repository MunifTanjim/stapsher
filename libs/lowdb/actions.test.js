const { getUsersCache } = require('../lowdb')

expect.addSnapshotSerializer({
  print: o => JSON.stringify(o, null, 2),
  test: o => o && typeof o === 'object'
})

const {
  createInstallationOnCache,
  deleteInstallationFromCache,
  addReposToCache,
  removeReposFromCache,
  fetchInstallationIdFromCache,
  addRepoToCache,
  incrementEntryCountCache
} = require('./actions')

let installation, repos, cache

/* the sequence of testing functions matters here */

describe('lowdb:actions', () => {
  describe('createInstallationOnCache', () => {
    it('works as expected', async () => {
      expect.assertions(2)

      installation = {
        id: 10,
        account: { id: 7, login: 'Harold', type: 'Admin' }
      }
      repos = [{ id: 1, name: 'TheMachine', full_name: 'Harold/TheMachine' }]
      await createInstallationOnCache(installation, repos)
      cache = await getUsersCache()
      expect(cache.getState()).toMatchSnapshot()

      installation = {
        id: 11,
        account: { id: 13, login: 'Arthur', type: 'Engineer' }
      }
      repos = [{ id: 666, name: 'Samaritan', full_name: 'Arthur/Samaritan' }]
      await createInstallationOnCache(installation, repos)
      cache = await getUsersCache()
      expect(cache.getState()).toMatchSnapshot()
    })
  })

  describe('addReposToCache', () => {
    it('works as expected', async () => {
      installation = {
        id: 10,
        account: { id: 7, login: 'Harold', type: 'Admin' }
      }
      repos = [
        { id: 0, name: 'GodMode', full_name: 'Harold/GodMode' },
        { id: 2027, name: 'If-Then-Else', full_name: 'Harold/If-Then-Else' }
      ]
      await addReposToCache(installation, repos)
      cache = await getUsersCache()
      expect(cache.getState()).toMatchSnapshot()
    })
  })

  describe('removeReposFromCache', () => {
    it('works as expected', async () => {
      installation = {
        id: 10,
        account: { id: 7, login: 'Harold', type: 'Admin' }
      }
      repos = [
        { id: 0, name: 'GodMode', full_name: 'Harold/GodMode' },
        { id: 2027, name: 'If-Then-Else', full_name: 'Harold/If-Then-Else' }
      ]
      await removeReposFromCache(installation, repos)
      cache = await getUsersCache()
      expect(cache.getState()).toMatchSnapshot()
    })
  })

  describe('deleteInstallationFromCache', () => {
    it('works as expected', async () => {
      installation = {
        id: 11,
        account: { id: 13, login: 'Arthur', type: 'Engineer' }
      }
      repos = []
      await deleteInstallationFromCache(installation, repos)
      cache = await getUsersCache()
      expect(cache.getState()).toMatchSnapshot()
    })
  })

  describe('fetchInstallationIdFromCache', () => {
    it('works as expected', async () => {
      expect(
        await fetchInstallationIdFromCache({
          username: 'Harold',
          repository: 'TheMachine'
        })
      ).toBe(10)

      expect(
        await fetchInstallationIdFromCache({
          username: 'Arthur',
          repository: 'Samaritan'
        })
      ).toBe(null)
    })
  })

  describe('addRepoToCache', () => {
    it('works as expected', async () => {
      await addRepoToCache({
        id: 666,
        name: 'Samaritan',
        full_name: 'Arthur/Samaritan',
        installation_id: 11,
        owner: 'Arthur',
        owner_id: 13
      })
      cache = await getUsersCache()
      expect(cache.getState()).toMatchSnapshot()
    })
  })

  describe('incrementEntryCountCache', () => {
    it('increments entry count', async () => {
      expect.assertions(3)
      expect(await incrementEntryCountCache()).toBe(1)
      expect(await incrementEntryCountCache()).toBe(2)
      expect(await incrementEntryCountCache()).toBe(3)
    })
  })
})
