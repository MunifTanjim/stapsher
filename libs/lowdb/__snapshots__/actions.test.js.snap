// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lowdb:actions addRepoToCache works as expected 1`] = `
{
  "Harold": {
    "repos": {
      "TheMachine": {
        "id": 1,
        "name": "TheMachine",
        "full_name": "Harold/TheMachine",
        "installation_id": 10,
        "owner": "Harold",
        "owner_id": 7
      }
    }
  },
  "Arthur": {
    "repos": {
      "Samaritan": {
        "id": 666,
        "name": "Samaritan",
        "full_name": "Arthur/Samaritan",
        "installation_id": 11,
        "owner": "Arthur",
        "owner_id": 13
      }
    }
  }
}
`;

exports[`lowdb:actions addReposToCache works as expected 1`] = `
{
  "Harold": {
    "repos": {
      "TheMachine": {
        "id": 1,
        "name": "TheMachine",
        "full_name": "Harold/TheMachine",
        "installation_id": 10,
        "owner": "Harold",
        "owner_id": 7
      },
      "GodMode": {
        "id": 0,
        "name": "GodMode",
        "full_name": "Harold/GodMode",
        "installation_id": 10,
        "owner": "Harold",
        "owner_id": 7
      },
      "If-Then-Else": {
        "id": 2027,
        "name": "If-Then-Else",
        "full_name": "Harold/If-Then-Else",
        "installation_id": 10,
        "owner": "Harold",
        "owner_id": 7
      }
    }
  },
  "Arthur": {
    "repos": {
      "Samaritan": {
        "id": 666,
        "name": "Samaritan",
        "full_name": "Arthur/Samaritan",
        "installation_id": 11,
        "owner": "Arthur",
        "owner_id": 13
      }
    }
  }
}
`;

exports[`lowdb:actions createInstallationOnCache works as expected 1`] = `
{
  "Harold": {
    "repos": {
      "TheMachine": {
        "id": 1,
        "name": "TheMachine",
        "full_name": "Harold/TheMachine",
        "installation_id": 10,
        "owner": "Harold",
        "owner_id": 7
      }
    }
  }
}
`;

exports[`lowdb:actions createInstallationOnCache works as expected 2`] = `
{
  "Harold": {
    "repos": {
      "TheMachine": {
        "id": 1,
        "name": "TheMachine",
        "full_name": "Harold/TheMachine",
        "installation_id": 10,
        "owner": "Harold",
        "owner_id": 7
      }
    }
  },
  "Arthur": {
    "repos": {
      "Samaritan": {
        "id": 666,
        "name": "Samaritan",
        "full_name": "Arthur/Samaritan",
        "installation_id": 11,
        "owner": "Arthur",
        "owner_id": 13
      }
    }
  }
}
`;

exports[`lowdb:actions deleteInstallationFromCache works as expected 1`] = `
{
  "Harold": {
    "repos": {
      "TheMachine": {
        "id": 1,
        "name": "TheMachine",
        "full_name": "Harold/TheMachine",
        "installation_id": 10,
        "owner": "Harold",
        "owner_id": 7
      }
    }
  }
}
`;

exports[`lowdb:actions removeReposFromCache works as expected 1`] = `
{
  "Harold": {
    "repos": {
      "TheMachine": {
        "id": 1,
        "name": "TheMachine",
        "full_name": "Harold/TheMachine",
        "installation_id": 10,
        "owner": "Harold",
        "owner_id": 7
      }
    }
  },
  "Arthur": {
    "repos": {
      "Samaritan": {
        "id": 666,
        "name": "Samaritan",
        "full_name": "Arthur/Samaritan",
        "installation_id": 11,
        "owner": "Arthur",
        "owner_id": 13
      }
    }
  }
}
`;
